### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by DenesDanko**
Read, taste, build, comment and approve It!

I can compile and build the page, so you should do too.
Since It is TS migration only, I do not expect new bugs, but feel free to test as well.

After 2 Approve, I merge the branch to the master.
Thus anyone can join the TS challenge.

### CODE REVIEW

**Comment by KristofNalesnyik** [comment]
It is better to use lowercase boolean, than uppercase Boolean, because the latter is an object, and I think it could cause unexpected behavior. https://www.tutorialsteacher.com/typescript/typescript-boolean
Actually it maybe only occurs in ComingSoon.tsx :)

File: client/src/components/ComingSoon/ComingSoon.tsx [hunk 1]
```diff
@@ -2,9 +2,14 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import './ComingSoon.scss';
 
-const ComingSoon = (props) => {
+interface ComingSoonProps {
+    pageName: string,
+    isPlural: Boolean
```
Comment (KristofNalesnyik) [inline client/src/components/ComingSoon/ComingSoon.tsx]: here.

File: client/src/components/LanguageSelector/LanguageSelector.tsx [hunk 2]
```diff
@@ -3,18 +3,20 @@ import { UIcontext } from '../contexts/UIcontext/UIcontext';
 import './LanguageSelector.scss';
 import PropTypes from 'prop-types';
 
-const LanguageSelector = (props) => {
-    const defaultLang = localStorage.getItem('lang');
-    const [sliderPos, setSliderPos] = useState(defaultLang === 'hu' ? 'left' : 'right');
+interface LanguageSelectorProps {
+    size?: 'small'
```
Comment (KristofNalesnyik) [inline client/src/components/LanguageSelector/LanguageSelector.tsx:7]: Does it mean that "size" has a default value "small"?
I dont know too much about TS interfaces, but
--> I would write here that type of 'size' is an enum, and in line 10, that default value of size is small.

File: client/src/components/MemberDetails/MemberDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    selectedMemberData: RegisteredUserType
+};
+
+const MemberDetails: React.FC<MemberDetailsProps> = ({ closeDialog, selectedMemberData: member }) => {
+    const [dataCopy, setDataCopy] = useState(false);
+    const { modalMembersDictionary, generalTermsDictionary, personalPagePlaceholders } = useContext(UIcontext).dictionary;
+    const { COPY, COPYTOCLIPBOARD } = modalMembersDictionary;
+    const { memberDetails } = useContext(UIcontext).dictionary;
+    const { SPIRITUALNAME, DATEOFBIRTH, GENDER, LEVELOFSTUDY, EMAIL, MOBILE, ADDRESS, EMNAME, EMMOBILE, EMEMAIL } = memberDetails;
+    const { modalButtons } = useContext(UIcontext).dictionary;
```
Comment (KristofNalesnyik) [inline client/src/components/MemberDetails/MemberDetails.tsx:20]: lines 18 & 20 could be written in 1 line with 16.
I think you know that.
Is it more readable / understandable the way you composed it?

File: package.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
   "description": "This package exists only to hold eslint and It's dependency",
   "main": "n.a",
   "scripts": {
-    "dev": "sed -i 's/DEV_SERVER.*/DEV_SERVER = 1/' server/.env.atlas; concurrently \"cd client && npm start\" \"cd server && npm run atlas\"",
+    "dev": "sed -i 's/DEV_SERVER.*/DEV_SERVER = 1/' server/.env.atlas; concurrently -s --kill-others --names \"BUILD,SERVER\" -c \"bgBlue,bgGreen\" \"cd client && npm start\" \"cd server && npm run atlas\"",
```
Comment (KristofNalesnyik) [inline package.json:7]: This line is a bit brutal. & what does it mean?

File: client/src/components/ComingSoon/ComingSoon.tsx (hunk 1 above)
Comment (DenesDanko) [reply client/src/components/ComingSoon/ComingSoon.tsx]: Good catch!

File: client/src/components/LanguageSelector/LanguageSelector.tsx (hunk 2 above)
Comment (DenesDanko) [reply client/src/components/LanguageSelector/LanguageSelector.tsx:7]: Enum would also work for sure.
I decided to use string union in many places because sometimes we display/change (like toUppercase)  the value on the interface. But not this one for sure.

File: client/src/components/MemberDetails/MemberDetails.tsx (hunk 3 above)
Comment (DenesDanko) [reply client/src/components/MemberDetails/MemberDetails.tsx:20]: Copy&Paste original code. ;o)

File: package.json (hunk 4 above)
Comment (DenesDanko) [reply package.json:7]: sed: edit .env.atlas config file, and set DEV_SERVER value to 1.
concurrently: start 2 parallel process in the terminal
   - frontend dev server, logs with a blue [ BUILD ] tag,
   - backend server, logs with a green [ SERVER ] tag.

File: client/src/App.tsx [hunk 5]
```diff
@@ -9,17 +9,25 @@ import Loading from './pages/Loading/Loading';
...
 import { UIcontext } from './components/contexts/UIcontext/UIcontext';
+import { DataContext } from './components/contexts/DataContext/DataContext';
 import { dictionaryList } from './languages/dictionaryList';
 
-class App extends Component {
-    constructor (props) {
+interface Props {};
+
+type AppState = {
+    uiContext: UiContextType
```
Comment (dizsmek) [inline client/src/App.tsx]:
```suggestion
    uiContext: UiContextType;
```

File: client/src/App.tsx [hunk 6]
```diff
@@ -9,17 +9,25 @@ import Loading from './pages/Loading/Loading';
...
+};
+
+class App extends Component<Props, AppState> {
+    constructor (props: Props) {
         super(props);
 
         if (!localStorage.getItem('lang')) {
             localStorage.setItem('lang', 'hu');
         }
+        const _lang = localStorage.getItem('lang') as LANGUAGE;
```
Comment (dizsmek) [inline client/src/App.tsx:30]: I get a parsing error saying `Parsing error: Unexpected token, expected ";"` at the "as" keyword. I couldn't figure out why.

File: package.json (hunk 4 above)
Comment (KristofNalesnyik) [reply package.json:7]: thanks :)

File: client/src/pages/Members/Members.tsx [hunk 7]
```diff
@@ -9,40 +9,43 @@ import MemberDetails from '../../components/MemberDetails/MemberDetails';
...
 import { Button } from 'react-bootstrap';
 
-const Members = (props) => {
+//TODO: REMOVE ALL @ts-ignore comments
+
+const Members: React.FC<RouteComponentProps<{}>> = ({ history, location }) => {
     const { SHOWINGRESULTSFOR, SHOWALLMEMBERSBUTTON, SHOWINGMEMBERSTITLE } = useContext(UIcontext).dictionary.membersPage;
 
-    const [members, setMembers] = useState([]);
+    const [members, setMembers] = useState<RegisteredUserType[]>([]);
```
Comment (KristofNalesnyik) [inline client/src/pages/Members/Members.tsx:17]: Isn't it necessary to import type(s)? That's a convenient solution.

File: client/src/types/DictionaryType.d.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+type DictionaryType = {
+    [k:string]:{
+        [l:string]: string
+    }
+};
```
Comment (KristofNalesnyik) [inline client/src/types/DictionaryType.d.ts:5]: I dont really get it, but looks beautiful.

**Review by KristofNalesnyik — DISMISSED** [review]
i do not understand everything, cause by my lack of knowledge, e.g. DictionaryType & history is not clear for me, and that data types are not needed to be imported.
Checked the app manually, I approve the PR.

File: client/src/components/ComingSoon/ComingSoon.tsx (hunk 1 above)
Comment (KristofNalesnyik) [reply client/src/components/ComingSoon/ComingSoon.tsx]: Fix it, & I approve the PR

File: client/src/components/Alert/Alert.tsx [hunk 9]
```diff
@@ -32,7 +37,7 @@ function Alert (props) {
 Alert.propTypes = {
     alertClose: PropTypes.func.isRequired,
     alertMsg: PropTypes.string,
-    alertType: PropTypes.string.isRequired
+    alertType: PropTypes.oneOf(['NOALERT', 'ERROR', 'WARNING', 'INFO'] as const).isRequired
```
Comment (dizsmek) [inline client/src/components/Alert/Alert.tsx:40]: What is the purpose of `as const` here?

File: client/src/components/Form/GenericDialog/GenericDialog.tsx [hunk 10]
```diff
@@ -39,7 +47,6 @@ GenericDialog.propTypes = {
     subtitle: PropTypes.string,
     reject: PropTypes.string,
     accept: PropTypes.string,
-    children: PropTypes.node.isRequired,
```
Comment (dizsmek) [inline client/src/components/Form/GenericDialog/GenericDialog.tsx:42]: Why did you remove the `children` props? And why isn't it in the `GenericDialogProps` interface?

File: client/src/components/Form/InputAvatar/InputAvatar.tsx [hunk 11]
```diff
@@ -5,8 +5,14 @@ import './InputAvatar.scss';
...
 import { Row, Col } from 'react-bootstrap';
+
+interface InputAvatarProps {
+    profileImgURL: string,
+    updateProfileImg: (event: React.ChangeEvent<HTMLInputElement>) => void
+};
+
 // user can't delete the photo - we might want to leave it like that; photo is mandatory
-const InputAvatar = (props) => {
+const InputAvatar: React.FC<InputAvatarProps> = (props) => {
```
Comment (dizsmek) [inline client/src/components/Form/InputAvatar/InputAvatar.tsx:15]: Why not write get the props like so: `({profileImgURL, updateProfileImg}) => {...}` instead of just getting the props and extracting its properties later?

File: client/src/components/Header/Header.js [hunk 12]
```diff
@@ -183,7 +183,7 @@ const Header = (props) => {
                             className="d-none d-sm-flex"
                         />
                         <Figure.Caption className={`avatar-name d-none ${searching ? '' : 'd-sm-flex'}`} as='h2'>
-                            <ActiveUserNameWrapper />
+                            { ActiveUserNameWrapper() }
```
Comment (dizsmek) [inline client/src/components/Header/Header.js:186]: If it's now a function returning plain text, I think we should rename it to something like `getActiveUserName`.

File: client/src/components/Header/Header.js [hunk 13]
```diff
@@ -220,12 +220,7 @@ const Header = (props) => {
                                             <li key={key} onClick={() => { handleSearchResultClick(user._id); }}>
                                                 <p>
                                                     { user.spiritualName !== '-' && <span>{user.spiritualName}</span> }
-                                                    <span>
-                                                        <AnyUserNameWrapper
-                                                            firstName={user.firstName}
-                                                            lastName={user.lastName}
-                                                        />
-                                                    </span>
+                                                    <span>{ AnyUserNameWrapper(user.firstName, user.lastName) }</span>
```
Comment (dizsmek) [inline client/src/components/Header/Header.js:223]: Same idea here, this could be `internationalizeName`.

File: client/src/pages/Members/Members.tsx (hunk 7 above)
Comment (DenesDanko) [reply client/src/pages/Members/Members.tsx:17]: Excellent observation!
NO type imports at all! ;o)
And the reason: files with '...name.**_d_**.ts' are imported automagically!

File: client/src/types/DictionaryType.d.ts (hunk 8 above)
Comment (DenesDanko) [reply client/src/types/DictionaryType.d.ts:5]: This type describe an object, where the 'k' key is string ( we do NOT specify the NUMBER of 'k' keys! )
K's value is another object, where the key 'l' is a string, and l's value is also a string ( we do not specify the NUMBER of 'l' keys!)

The reason:
We do not know at all how many 'k' and 'l' keys do we have in the language dictionaries!
The specification above describe a GENERIC form of the dictionaries, w/o the need of specify the number of 'k'-s or 'l'-s!

This means, this is a valid Dict. type:
{
    "transactionTypes": {
            "PAYMENT": "payment",
            "DEBT": "debt"
    },
    "addTransactionDialog": {
            "BEFORETYPE": "Add ",
            "AFTERTYPE": "",
            "DATESELECTORINFO": "click on the field to select date"
    }
}
Here we have 2x'k' and 2x'l' and 3x'l'. So extend the Dictinary is easy.

**Review by KristofNalesnyik — APPROVED** [review]

File: client/src/components/LanguageSelector/LanguageSelector.tsx [hunk 14]
```diff
@@ -26,15 +28,15 @@ const LanguageSelector = (props) => {
 
     return (
 
-        <div className = {`language-selector ${props.size ? props.size : ''}`} onClick = {handleSliderClick}>
+        <div className = {`language-selector ${size ? size : ''}`} onClick = {handleSliderClick}>
```
Comment (dizsmek) [inline client/src/components/LanguageSelector/LanguageSelector.tsx]: The ternary in this line is unnecessary, because if size is undefined, the div won't get any class beside `"language-selector"`.
```suggestion
        <div className = {`language-selector ${size}`} onClick = {handleSliderClick}>
```

**Comment by rigo101** [comment]
Itt vannak egyben a komentjeim:

Alert:
alertType: PropTypes.oneOf(['NOALERT', 'ERROR', 'WARNING', 'INFO'] as const).isRequired- nekem alahuzza az as-t:
	Parsing error: Unexpected token, expected "," 
Ilyen van a MemberDetails-ben, a Members-ben es az App-ban is.

GenericDialog:
ha a PropTypes-ba beletettem a children: PropTypes.node-ot, 
es az interface-be pedig a children: React.ReactNode-t, es csak ekkor nem kaptam hibat a children-re

Logout:
event: React.MouseEvent<HTMLButtonElement, MouseEvent>):void
Ez miert mouse event? Es miert kell 2szer beirni?
Ilyen van a MemberDetails-ben, es a Registration-be is.

MemberDetails: 
miert kell a props-oknal a members?: selectedMemberData: member 

RegisteredUserType
az elso 3 key miert string?

Registration:
miert type es miert nem interface van itt hasznalva?

96 sor: alahuzza a files-t
(property) HTMLInputElement.files: FileList | null
Returns a FileList object on a file type input object.
Parsing error: Unexpected token, expected ";"

File: client/src/types/RegisteredUserType.d.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+    birthday?: string | null,
+    gender?: GENDER | null,
+    level?: LEVEL | null,
+    email?: string | null,
+    mobile?: string | null,
+    address?: string | null,
+    emName?: string | null,
+    emMobile?: string | null,
+    emEmail?: string | null
+};
```
Comment (dizsmek) [inline client/src/types/RegisteredUserType.d.ts]: Why are some properties in quotation marks and others not?

File: client/src/types/UiContextType.d.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+type UiContextType = {
+    isHamburgerOpen: boolean,
+    toggleHamburger: () => void,
+    closeHamburger: () => void,
+
+    isSuperuser: boolean,
+    isFinanceAdmin: boolean,
+    isEventAdmin: boolean,
+    isYogaAdmin: boolean,
+    setAccess: (a: boolean, b: boolean, c: boolean, d: boolean) => void,
```
Comment (dizsmek) [inline client/src/types/UiContextType.d.ts:10]: Itt nem mukodik valami olyan, hogy `(...a: boolean[]) => void` ahelyett, hogy felsoroljuk a parametereket?

**Review by dizsmek — APPROVED** [review]

**Comment by DenesDanko** [comment]
Alert: ez v.mi kornyezeti beallitas lesz.
    Mi a ts es babel verzioja azoknal akik latjak a hibat? Nalam ez OK ui, es a verzioim:
    $ tsc --version
    Version 4.1.3
    $ babel --version
    6.26.0 (babel-core 6.26.3)
    
GenericDialog: javitom

Logout: a MouseEvent beepitett React type, 2 parametert var.
Viszont itt nem hasznaljuk semmire az 'event' parametert, ezert ki fogom venni.
A masik ket hely OK igy.

MemberDetails: ez a syntax atnevezi a props-ot. Hasonlo van a PrivateRoute-ban is (component: Component)

RegisteredUserType: nyomdahuba lehet, javitom

Registration: a class component-ek 2 tipust varnak a <> kozott. Az elso a props, a masodik a state:
    class Registration extends Component<RouteComponentProps<{}>, RegistrationState>
    A props mindig kivulrol jon, itt a tipus RouteComponentProps, es mivel nem varunk props-ot ezert empty object {}
    A state NEM kulso parameter, ezert gondoltam hogy a TYPE jobban passzol ide. Hogy miert nincs a tobbi type mellett?
    Azert mert ezt a tipust -velhetoen- csak itt fogjuk hasznalni.. Ha mashol is lenne, ki lehetne/kellene vinni..

    96. sor: lasd Alert valasz a comment tetejen.

File: client/src/App.tsx (hunk 5 above)
Comment (DenesDanko) [reply client/src/App.tsx]: Fixed in: 3fae08340f1a3189346caba72e9c71ca08916be0

File: client/src/App.tsx (hunk 6 above)
Comment (DenesDanko) [reply client/src/App.tsx:30]: Please check my answer to Ildi's comment. Probably environment problem. Need further investigation.

File: client/src/components/Alert/Alert.tsx (hunk 9 above)
Comment (DenesDanko) [reply client/src/components/Alert/Alert.tsx:40]: Type constraints I think. W/o 'as const' PropTypes validation shows me errors.

File: client/src/components/Form/GenericDialog/GenericDialog.tsx (hunk 10 above)
Comment (DenesDanko) [reply client/src/components/Form/GenericDialog/GenericDialog.tsx:42]: Fixed in: 3fae083
But .isRequired missing..Damn! Fix coming soon!

File: client/src/components/Form/InputAvatar/InputAvatar.tsx (hunk 11 above)
Comment (DenesDanko) [reply client/src/components/Form/InputAvatar/InputAvatar.tsx:15]: I add TS related code and did NOT refactor legacy code, to separate TS issues and refactor problems.
But we can follow this principle in the future.

File: client/src/components/Header/Header.js (hunk 12 above)
Comment (DenesDanko) [reply client/src/components/Header/Header.js:186]: I would follow It in a ticket, since It is refactor and not TS related...

File: client/src/components/Header/Header.js (hunk 13 above)
Comment (DenesDanko) [reply client/src/components/Header/Header.js:223]: Same answere here ;o)
I would follow It in a ticket, since It is refactor and not TS related...

File: client/src/components/LanguageSelector/LanguageSelector.tsx (hunk 14 above)
Comment (DenesDanko) [reply client/src/components/LanguageSelector/LanguageSelector.tsx]: Fixed in: 3fae083

File: client/src/types/RegisteredUserType.d.ts (hunk 15 above)
Comment (DenesDanko) [reply client/src/types/RegisteredUserType.d.ts]: Fixed in: 3fae083

File: client/src/types/UiContextType.d.ts (hunk 16 above)
Comment (DenesDanko) [reply client/src/types/UiContextType.d.ts:10]: I am not sure!
In TS parameter names are just placeholders. You can find in our code that I use parameter names like 'a', 'b' in type definitions. But in real functions they are 'firstName' and 'lastName'..
So the POSITION is important and not the names.
